iree/compiler/Dialect/Flow/Transforms/InferNumericNarrowing.cpp

```diff
@@ -0,0 +1,142 @@
+// Copyright 2021 The IREE Authors
+//
+// Licensed under the Apache License v2.0 with LLVM Exceptions.
+// See https://llvm.org/LICENSE.txt for license information.
+// SPDX-License-Identifier: Apache-2.0 WITH LLVM-exception
+
+#include "iree/compiler/Dialect/Flow/Transforms/PassDetail.h"
+#include "iree/compiler/Dialect/Flow/Transforms/Passes.h"
+#include "iree/compiler/Dialect/Util/Analysis/Attributes/Range.h"
+#include "iree/compiler/Dialect/Util/Analysis/DFX/Solver.h"
+#include "iree/compiler/Dialect/Util/Analysis/DFX/State.h"
+#include "iree/compiler/Dialect/Util/Analysis/Explorer.h"
+#include "iree/compiler/Dialect/Util/IR/UtilDialect.h"
+#include "iree/compiler/Dialect/Util/IR/UtilOps.h"
+#include "llvm/ADT/SmallPtrSet.h"
+#include "llvm/Support/Debug.h"
+#include "mlir/Dialect/Linalg/IR/LinalgInterfaces.h"
+
+using llvm::SmallPtrSet;
+
+namespace mlir {
+namespace iree_compiler {
+namespace IREE {
+namespace Flow {
+
+namespace {
+
+IntegerType deriveIntegerTypeFromRange(MLIRContext *context, int64_t minValue,
+                                       int64_t maxValue) {
+  // Clamp min/max to span 0.
+  const int64_t zero = 0;
+  minValue = std::min(zero, minValue);
+  maxValue = std::max(zero, maxValue);
+  bool isSigned;
+  if (minValue < 0) {
+    // For signed, make symmetric from -N:N-1
+    isSigned = true;
+    maxValue = std::max(std::abs(minValue) - 1, maxValue);
+    minValue = std::min(-maxValue - 1, minValue);
+  } else {
+    isSigned = false;
+  }
+  int64_t n = maxValue - minValue + 1;
+  int64_t numBits = std::ceil(std::log2(n));
+
+  return IntegerType::get(context, numBits,
+                          isSigned
+                              ? IntegerType::SignednessSemantics::Signed
+                              : IntegerType::SignednessSemantics::Unsigned);
+}
+
+class InferNumericNarrowingPass
+    : public InferNumericNarrowingBase<InferNumericNarrowingPass> {
+  void getDependentDialects(DialectRegistry &registry) const override {
```

Comment:
I think this is something you can declare in the passes.td:
`let dependentDialects = ["IREE::Util::UtilDialect"];`
(though we have a lot of passes to clean up to use that, so fine like this here)

Reply:
I think that is of dubious value and potentially damaging for these cases: Because the tablegen stuff emits that in the detail header, it requires everyone to pull in the headers for the dependent dialects. Aside from being a waste, it creates more iwyu slop since now every pass will have that implicitly.